libraries/Networking/src/ESP_Network_Interface.cpp

```diff
@@ -0,0 +1,526 @@
+{
+    if(!mac || _esp_netif == NULL){
+        return NULL;
+    }
+    esp_err_t err = esp_netif_get_mac(_esp_netif, mac);
+    if(err != ESP_OK){
+        log_e("Failed to get netif mac: %d", err);
+        return NULL;
+    }
+    // getMac(mac);
+    return mac;
+}
+
+String ESP_Network_Interface::macAddress(void)
+{
+    uint8_t mac[6] = {0,0,0,0,0,0};
+    char macStr[18] = { 0 };
+    macAddress(mac);
+    sprintf(macStr, "%02X:%02X:%02X:%02X:%02X:%02X", mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]);
+    return String(macStr);
+}
+
+IPAddress ESP_Network_Interface::localIP()
+{
+    if(_esp_netif == NULL){
+        return IPAddress();
+    }
+    esp_netif_ip_info_t ip;
+    if(esp_netif_get_ip_info(_esp_netif, &ip)){
+        return IPAddress();
+    }
+    return IPAddress(ip.ip.addr);
+}
+
+IPAddress ESP_Network_Interface::subnetMask()
+{
+    if(_esp_netif == NULL){
+        return IPAddress();
+    }
+    esp_netif_ip_info_t ip;
+    if(esp_netif_get_ip_info(_esp_netif, &ip)){
+        return IPAddress();
+    }
+    return IPAddress(ip.netmask.addr);
+}
+
+IPAddress ESP_Network_Interface::gatewayIP()
+{
+    if(_esp_netif == NULL){
+        return IPAddress();
+    }
+    esp_netif_ip_info_t ip;
+    if(esp_netif_get_ip_info(_esp_netif, &ip)){
+        return IPAddress();
+    }
+    return IPAddress(ip.gw.addr);
+}
+
+IPAddress ESP_Network_Interface::dnsIP(uint8_t dns_no)
+{
+    if(_esp_netif == NULL){
+        return IPAddress();
+    }
+    esp_netif_dns_info_t d;
+    if(esp_netif_get_dns_info(_esp_netif, dns_no?ESP_NETIF_DNS_BACKUP:ESP_NETIF_DNS_MAIN, &d) != ESP_OK){
+        return IPAddress();
+    }
+    return IPAddress(d.ip.u_addr.ip4.addr);
+}
+
+IPAddress ESP_Network_Interface::broadcastIP()
+{
+    if(_esp_netif == NULL){
+        return IPAddress();
+    }
+    esp_netif_ip_info_t ip;
+    if(esp_netif_get_ip_info(_esp_netif, &ip)){
+        return IPAddress();
+    }
+    return calculateBroadcast(IPAddress(ip.gw.addr), IPAddress(ip.netmask.addr));
+}
+
+IPAddress ESP_Network_Interface::networkID()
+{
+    if(_esp_netif == NULL){
+        return IPAddress();
+    }
+    esp_netif_ip_info_t ip;
+    if(esp_netif_get_ip_info(_esp_netif, &ip)){
+        return IPAddress();
+    }
+    return calculateNetworkID(IPAddress(ip.gw.addr), IPAddress(ip.netmask.addr));
+}
+
+uint8_t ESP_Network_Interface::subnetCIDR()
+{
+    if(_esp_netif == NULL){
+        return (uint8_t)0;
+    }
+    esp_netif_ip_info_t ip;
+    if(esp_netif_get_ip_info(_esp_netif, &ip)){
+        return (uint8_t)0;
+    }
+    return calculateSubnetCIDR(IPAddress(ip.netmask.addr));
+}
+
+IPAddress ESP_Network_Interface::localIPv6()
+{
+    if(_esp_netif == NULL){
+        return IPAddress(IPv6);
+    }
+    static esp_ip6_addr_t addr;
+    if(esp_netif_get_ip6_linklocal(_esp_netif, &addr)){
+        return IPAddress(IPv6);
+    }
+    return IPAddress(IPv6, (const uint8_t *)addr.addr, addr.zone);
+}
+
+IPAddress ESP_Network_Interface::globalIPv6()
```

Comment:
Be consistent with IPv4.

The `localIPv4()` function returns a local (as opposed to remote) IPv4 address with global scope.

This function returns a local (as opposed to remote) IPv6 address with global scope -- the same thing, but IPv6 instead of IPv4, so it should have a similar name.

Reply:
`linkLocalIPv6()` is a win for me. I like it!

Almost all cases will be either:
* 1x Public IPv6, 1x IPv4 (probably a private range), 1x Link-Local IPv6; or
* 1x IPv4, 1x Link-Local IPv6 (if no IPv6 available)

It would be rare to have extra IPv6 as ESP / LWIP only supports RA prefixes; it does not support DHCPv6 stateful, or privacy extensions, so won't have additional addresses that a laptop / tablet / phone might.

Less common scenarios:
* 1x ULA IPv6, either with or without a public IPv6; this could be from someone wanting to test out local IPv6 but not supported by their ISP (no public), or want a private network (similar to IPv4 private address). Not a lot of commercial implementation, although a Thread gateway may use this for the Thread network. May also be present in NAT66 scenarios
* 1x static ULA IPv6 may be used in AP scenarios (possibly 2x with a simple address fd00::1, plus a EUI-64 address)
* Multiple public IPv6: unlikely on one interface. Could be static configuration if you want a simple fixed address plus RA address.
* Multiple link-local IPv6: could be static configuration if you want a simple fixed address (fe80::1), plus an RA address (although ULA are more accessible, e.g. via browser)

Reply:
`linkLocalIPv6()` aligns with what is available in IDF / LWIP, i.e. `esp_netif_get_ip6_linklocal()`

The initial `globalIPv6()` aligns with `esp_netif_get_ip6_global()`, which I think only returns a 2xxx / 3xxx address.

It would be rare to have multiple public (what ESP calls GLOBAL, i.e. 2xxx or 3xxx) addresses, so I don't think `globalIPv6(int index)` is that useful as there is unlikely more than one. I think it would be more common to have a ULA instead, although if you mean global scope then technically that includes ULAs (but not in ESP). 

Was the index supposed to align with the ` esp_netif_get_all_ip6()` function?  This covers the ULA scenario (single and multiple) as well as the multiple link-local (not very useful for a device I think), as well as where you actually have multiple 2xxx / 3xxx addresses (e.g. one static plus one from RA).

If it did align with the underlying function then maybe `allIPv6(index=0)` would be a better name, as it could return global scope 2xxx/3xxx, ULA, and/or link-local.

Reply:
I have two global addresses on my ESP's ETH port, so the is why I proposed the index. And yes, it will come from `esp_netif_get_all_ip6()`